### PR TITLE
Add `ListAddresses`, `PutWallet`, `PutWalletPassphrase` to the Servant API definition.

### PR DESCRIPTION
**Pull request by jonathanknowles**
# Issue Number

Issue #118 

# Overview

- [x] Add `ListAddresses` operation to the Servant API definition.
- [x] Add `PutWallet` operation to the Servant API definition.
- [x] Add `PutWalletPassphrase` operation to the Servant API definition.

### CODE REVIEW

**Comment by KtorZ** [comment]
Arg! We agreed on the scope during the last meeting:

- getWallet
- listWallet
- postWallet
- postTransaction
- listTransaction

I've put the ticket in QA already; let's get these merged, but we can move on to something else now :+1:

File: test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 1]
```diff
@@ -348,25 +364,27 @@ specification =
...
+
+-- | Utility function to provide an ad-hoc 'ToSchema' instance for a definition:
+-- we simply look it up within the Swagger specification.
+declareSchemaForDefinition :: T.Text -> Declare (Definitions Schema) NamedSchema
+declareSchemaForDefinition name =
+    case specification ^. definitions . at name of
+        Nothing -> error $
+            "unable to find the definition for " <> show name <> " in the spec"
+        Just schema ->
+            return $ NamedSchema (Just name) schema
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/Api/TypesSpec.hs:412]: I think I'd even push it further and make it all "automagic" via a `Typeable` instance :thinking:  ?

File: src/Cardano/Wallet/Api.hs [hunk 2]
```diff
@@ -24,8 +42,11 @@ type Api = Wallets
 type Wallets =
     DeleteWallet
     :<|> GetWallet
+    :<|> ListAddresses
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api.hs]: The idea was to have groups matching the groups from the API specification :sweat_smile: 
So I'd suggest to have:

```hs
type Wallets =
    DeleteWallet
    :<|> GetWallet
    :<|> ListWallets
    :<|> PostWallet
    :<|> PutWallet
    :<|> PutWalletPassphrase

type Addresses =
    listAddresses
```

File: src/Cardano/Wallet/Api/Types.hs [hunk 3]
```diff
@@ -101,6 +108,14 @@ import qualified Data.Text.Encoding as T
                                   API Types
 -------------------------------------------------------------------------------}
 
+data Address = Address
+    { _id :: !(ApiT P.Address)
+    , _state :: !AddressState
+    } deriving (Eq, Generic, Show)
+
+data AddressState = Used | Unused
+    deriving (Eq, Generic, Show)
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api/Types.hs]: That, we want in our `Primitive.Model` the same way we have other metadata about various core types.

File: src/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -164,6 +188,27 @@ newtype ApiMnemonicT (sizes :: [Nat]) (purpose :: Symbol) =
...
+
+instance FromJSON AddressState where
+    parseJSON = genericParseJSON defaultSumTypeOptions
+instance ToJSON AddressState where
+    toJSON = genericToJSON defaultSumTypeOptions
+
+instance FromJSON (ApiT P.Address) where
+    parseJSON =
+        parseJSON >=>
+          maybe
```
Comment (KtorZ) [inline src/Cardano/Wallet/Api/Types.hs]: I'd drop the pointfree here and use a `do-notation` to makes the reading a bit clearer.

File: src/Cardano/Wallet/Api.hs (hunk 2 above)
Comment (jonathanknowles) [reply src/Cardano/Wallet/Api.hs]: Interesting! I was under the impression that the `listAddresses` operation was within the `wallets` group:

```
  /wallets/{walletId}/addresses:                                                                                                                                                      
    get:                                                                                                                                                                              
      operationId: listAddresses      
```
Should we update the specification to move it to another group?

File: test/data/Cardano/Wallet/Api/Address.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        },
+        {
+            "state": "used",
+            "id": "UT4QpKUSxauQo1XQ6VoT8MhVPYPn5m8tPMgE5WHnLy8E3X1ete7NAiXqenUtoGLKGo9P"
+        },
+        {
+            "state": "unused",
+            "id": "Kgihnukc1UStfMcKxBjRew4jnt7Z53JAUiKZkw3XErXy9NTxZNfmVarXhP9rs99PDFFA"
+        }
+    ]
```
Comment (KtorZ) [inline test/data/Cardano/Wallet/Api/Address.json:20]: I really like we have golden tests for all types. It helps controlling the shape of the JSON objects right away :ok_hand:

File: test/data/Cardano/Wallet/Api/WalletPutData.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+    "seed": -3282480895854715413,
+    "samples": [
+        {},
+        {},
+        {},
+        {
+            "name": "ro8n&(V$]%dJ吚 (v]QSu1𩈾DfZ|*w?\\Iy,X%d9k}[𦴽YFJtW?7ZJv_-jYz_KpB`iE@$*4aV)`𠪚V`Q&0O7=ZmB<KG3Y5$\\>WO$1(/4子tmkNd;Zcte𥗼fay>nqh𨃮wQwW~u:`,pzhSB6|&~p}\"dS6🢓_3#*+f#5𠷓>:-Yq58Q_:𠷒as0/^j7;^BafT3I3y䖜jY0<CiwxqJ.Ofx<\" 1FDU9HrlD$7𨸴cEIq+*>~"
```
Comment (KtorZ) [inline test/data/Cardano/Wallet/Api/WalletPutData.json:8]: That's the kind of name I imagine @piotr-iohk would use on a wallet yeah :no_mouth:

File: src/Cardano/Wallet/Api.hs (hunk 2 above)
Comment (KtorZ) [reply src/Cardano/Wallet/Api.hs]: Nope, the spec is correct here. Endpoints are grouped by `tags`, cf:

https://github.com/input-output-hk/cardano-wallet/blob/master/specifications/api/swagger.yaml#L896

https://github.com/input-output-hk/cardano-wallet/blob/master/specifications/api/swagger.yaml#L844

So, even if, conceptually, the `Address` resource is a sub-resource of the `Wallet` resource, operations related to addresses are in a separate group to be easier to identify (same apply for transactions).

**Comment by jonathanknowles** [comment]
> Arg! We agreed on the scope during the last meeting:
>     * getWallet
>     * listWallet
>     * postWallet
>     * postTransaction
>     * listTransaction

I had interpreted the scope to be everything within issue #53, which states:
> Extend the API to support all endpoints.

But now I notice that it also states (later on):
> The scope of this ticket has been reduced down to the endpoints discussed during the last iteration 
planning meeting

In which case I now understand your comment. :)

In that case, perhaps we should update the the scope of issue #53, so that it no longer states "all endpoints"?

File: test/data/Cardano/Wallet/Api/Address.json (hunk 5 above)
Comment (jonathanknowles) [reply test/data/Cardano/Wallet/Api/Address.json:20]: :+1:

**Comment by KtorZ** [comment]
@jonathanknowles Yes, let's update the scope. I mean, no big deal, we'll to get them done at some point. But I'd integrate endpoints incrementally and start implementing handlers and CLI commands for those we already have, so that we are able to deliver _something_ already.

File: src/Cardano/Wallet/Api/Types.hs (hunk 4 above)
Comment (jonathanknowles) [reply src/Cardano/Wallet/Api/Types.hs]: Fair enough :)

**Comment by jonathanknowles** [comment]
> start implementing handlers and CLI commands for those we already have, so that we are able to deliver _something_ already.

Okay!

**Comment by jonathanknowles** [comment]
>  Yes, let's update the scope.

Done.

File: src/Cardano/Wallet/Api/Types.hs (hunk 3 above)
Comment (jonathanknowles) [reply src/Cardano/Wallet/Api/Types.hs]: Done.

File: src/Cardano/Wallet/Api.hs (hunk 2 above)
Comment (jonathanknowles) [reply src/Cardano/Wallet/Api.hs]: Okay! I've moved `ListAddresses` to a separate group.

**Review by KtorZ — APPROVED** [review]
LGTM.

I've rebased and squashed on top of latest `master` :+1:

File: test/data/Cardano/Wallet/Api/WalletPutData.json (hunk 6 above)
Comment (jonathanknowles) [reply test/data/Cardano/Wallet/Api/WalletPutData.json:8]: Buried deep within this name is the character "䖜", which apparently is "the fighting sound made by two tigers" (https://en.wiktionary.org/wiki/%E4%96%9C)

**Pull request merged**